lightning/src/routing/network_graph.rs

```diff
@@ -0,0 +1,1596 @@
+		write!(f, "src_node_id {}, last_update {}, enabled {}, cltv_expiry_delta {}, htlc_minimum_msat {}, fees {:?}", log_pubkey!(self.src_node_id), self.last_update, self.enabled, self.cltv_expiry_delta, self.htlc_minimum_msat, self.fees)?;
+		Ok(())
+	}
+}
+
+impl_writeable!(DirectionalChannelInfo, 0, {
+	src_node_id,
+	last_update,
+	enabled,
+	cltv_expiry_delta,
+	htlc_minimum_msat,
+	fees,
+	last_update_message
+});
+
+#[derive(PartialEq)]
+/// Details regarding a channel (both directions)
+pub struct ChannelInfo {
+	/// Protocol features of a channel communicated during its announcement
+	pub features: ChannelFeatures,
+	/// Details regarding one of the directions of a channel
+	pub one_to_two: DirectionalChannelInfo,
+	/// Details regarding another direction of a channel
+	pub two_to_one: DirectionalChannelInfo,
+	/// An initial announcement of the channel
+	//this is cached here so we can send out it later if required by initial routing sync
+	//keep an eye on this to see if the extra memory is a problem
+	pub announcement_message: Option<msgs::ChannelAnnouncement>,
+}
+
+impl std::fmt::Display for ChannelInfo {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+		write!(f, "features: {}, one_to_two: {}, two_to_one: {}", log_bytes!(self.features.encode()), self.one_to_two, self.two_to_one)?;
+		Ok(())
+	}
+}
+
+impl_writeable!(ChannelInfo, 0, {
+	features,
+	one_to_two,
+	two_to_one,
+	announcement_message
+});
+
+
+/// Fees for routing via a given channel or a node
+#[derive(Eq, PartialEq, Copy, Clone, Debug)]
+pub struct RoutingFees {
+	/// Flat routing fee
+	pub base_msat: u32,
+	/// Liquidity-based routing fee
+	pub proportional_millionths: u32,
+}
+
+impl Readable for RoutingFees{
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<RoutingFees, DecodeError> {
+		let base_msat: u32 = Readable::read(reader)?;
+		let proportional_millionths: u32 = Readable::read(reader)?;
+		Ok(RoutingFees {
+			base_msat,
+			proportional_millionths,
+		})
+	}
+}
+
+impl Writeable for RoutingFees {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		self.base_msat.write(writer)?;
+		self.proportional_millionths.write(writer)?;
+		Ok(())
+	}
+}
+
+
+#[derive(PartialEq)]
+/// Details regarding a node in the network
+pub struct NodeInfo {
+	/// All valid channels a node has announced
+	pub channels: Vec<u64>,
+	/// Lowest fees enabling routing via any of the known channels to a node
+	pub lowest_inbound_channel_fees: Option<RoutingFees>,
+	/// Protocol features the node announced support for
+	pub features: NodeFeatures,
+	/// When the last known update to the node state was issued
+	/// Unlike for channels, we may have a NodeInfo entry before having received a node_update.
+	/// Thus, we have to be able to capture "no update has been received", which we do with an
+	/// Option here.
+	pub last_update: Option<u32>,
+	/// Color assigned to the node
+	pub rgb: [u8; 3],
```

Comment:
Why arent these Options? Can we refactor all the stuff that shows up in the NodeAnnouncement into a separate struct and make it Option<>al here?

Reply:
Doing this for rgb and alias, not sure about features and addresses?

Because when we interpret features/addrs for None, we will indeed return Features::empty and empty vector, not just because it's a default value, but because that's the correct answer for a given call.

Reply:
For addresses, yea, I dont think so for features. There is  big difference between "dont know their feature set" and "they don't support anything".

Reply:
Oh, but actually, why not just make it one struct. This is a pretty classic example of "datastructures should capture what the reality of possible states are" - rgb, alias, addresses, features, and announcement_message should be in a separate sub-struct, eg the following. This super neatly captures, in both a human-understandable, as well as machine-understandable (ie compiler-enforced) way, that this set of fields is available only after the first announcement info, and after that, they are all available.

```
/// Information received in the latest node_announcement from this node.
pub struct NodeAnnouncementInfo {
       pub features: NodeFeatures,
       pub last_update: u32,
       pub rgb:[u8; 3],
       pub alias: [u8; 32],
       pub addresses: Vec<NetAddress>,
       pub announcement_message: Option<msgs::NodeAnnouncement>,
}
pub struct NodeInfo {
       pub channels: Vec<u64>,
       pub lowest_inbound_channel_fees: Option<RoutingFees>,
       pub announcement_info: Option<NodeAnnouncementInfo>
}
```